java-checks/src/main/java/org/sonar/java/checks/VolatileNonPrimitiveFieldCheck.java

```diff
@@ -85,11 +86,27 @@ public void visitNode(Tree tree) {
       .filter(m -> m.is(Tree.Kind.VARIABLE))
       .map(VariableTree.class::cast)
       .filter(v -> ModifiersUtils.hasModifier(v.modifiers(), Modifier.VOLATILE))
-      .filter(v -> !v.type().symbolType().isPrimitive())
-      .filter(v -> !isImmutableType(v.type().symbolType()))
+      .filter(v -> isUnsafeVolatile(v.type().symbolType()))
       .forEach(v -> reportIssue(ModifiersUtils.getModifier(v.modifiers(), Modifier.VOLATILE), v.type(), getMessage(v)));
   }
 
+  private static boolean isUnsafeVolatile(Type symbolType) {
+    return !(symbolType.isPrimitive()
+      || isImmutableType(symbolType)
+      || isSafelyAnnotated(symbolType.symbol().metadata()));
+  }
+
+  private static boolean isSafelyAnnotated(SymbolMetadata metadata) {
+    for (SymbolMetadata.AnnotationInstance a : metadata.annotations()) {
```

Comment:
what about writing this as stream instead?
```java
return metadata.annotations().stream()
  .map(SymbolMetadata.AnnotationInstance::symbol)
  .map(Symbol::type)
  .anyMatch(type -> type.is("javax.annotation.concurrent.Immutable") || type.is("javax.annotation.concurrent.ThreadSafe"));
```

Reply:
Looks better indeed.